### PR TITLE
feature: turbo knex

### PR DESCRIPTION
**Pull request by gjovs**
#### Description
This pull request enhances the cache functionality within Knex's select query results. Benchmark testing shows that integrating node-cache as the caching library reduces query execution time by 17x, while using MemJS results in an impressive ~50x reduction in time!

#### What type of PR contains? 
- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [x] 🔥 Performance Improvements
- [ ] 🧑‍💻 Code Refactor
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

### CODE REVIEW

File: src/infra/db/mssql/util/knex/extensions/turbo-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import k, { Knex } from 'knex';
+
+const ALLOWED_DRIVERS = ['MSSQL'];
+
+function getCurrentDrive(config: Knex.Config) {
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-plugin.ts:5]: This code is duplicated; why not create a utils file and import it?

File: src/infra/db/mssql/util/knex/extensions/turbo-plugin.ts
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-plugin.ts:1]: I liked the implementation!

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import { logger } from '@/util';
+import { makeCacheServer } from '@/infra/cache';
+
+type Services = 'memjs' | 'node-cache';
+type GenericObject = Record<string, unknown>;
+
+const cache = new NodeCache({ stdTTL: 100, checkperiod: 120 });
+const memCache = makeCacheServer();
+
+const MAX_MEMORY_USAGE = 100 * 1024 * 1024; // 100 MB
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:13]: This won't work. When running in production, this project consumes an average of 125MB under normal usage. For it to function correctly, your limit has to be at least 300MB.

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+async function getCachedQueryMemCache(sql: Knex.Sql) {
+  const key = cacheKey(sql);
+  const buffer = await memCache.get(key);
+
+  if (!buffer) return;
+
+  try {
+    return JSON.parse(buffer.toString());
+  } catch (error) {
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:50]: Rename error to _error

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    if (resultSize > MAX_RESULT_SIZE) {
+      logger.log({
+        level: 'warn',
+        message: `Result size ${resultSize} exceeds maximum allowed size of ${MAX_RESULT_SIZE}. Cache not saved.`
+      });
+      return;
+    }
+
+    await cacheQuery(sql, result, service);
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:77]: You don’t need that await; from what I can see, you don’t care whether the save was successful or not, and you’re not collecting results.

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  return cache.get(key);
+}
+
+async function getCachedQuery(sql: Knex.Sql, service: Services) {
+  if (service === 'node-cache') return getCachedQueryNodeCache(sql);
+  return getCachedQueryMemCache(sql);
+}
+
+function saveToCache(sql: Knex.Sql, result: unknown, service: Services) {
+  setImmediate(async () => {
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:66]: I was going to question the setImmediate, but upon closer analysis, it was well thought out. Congratulations!

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+          level: 'warn',
+          message: 'Memory usage exceeded maximum allowed. Cache not saved.'
+        });
+        return resolve(data);
+      }
+
+      saveToCache(sql, data, service);
+
+      resolve(data);
+    } catch (error) {
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:104]: Rename error to _error.

This is okay, I suspect this catch will never be triggered, but it's better to be safe...

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  };
+}
+
+export function turboInterceptorPlugin(knex: typeof k) {
+  const knexProxy = new Proxy(knex, {
+    apply(target, _, [arg]) {
+      const instanceOfKnex = target(arg);
+
+      const service: Services =
```
Comment (lucassm02) [inline src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:119]: A variable called DEFAULT_CACHE_SERVICE wouldn’t hurt.

**Review by lucassm02 — CHANGES_REQUESTED** [review]
Excellent work, it was a great implementation!!! 

If you have any questions about what I mentioned regarding the setImmediate or the try...catch, check out this [video](https://www.youtube.com/watch?v=eiC58R16hb8).

I am eagerly awaiting your adjustments.

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts (hunk 4 above)
Comment (gjovs) [reply src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:77]: I'll keep this to be able to wrap a try catch block inside this part of code to get a log inside the catch block if error throws

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts (hunk 2 above)
Comment (gjovs) [reply src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:13]: thanks for the context, I'll update the value

File: src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts (hunk 5 above)
Comment (gjovs) [reply src/infra/db/mssql/util/knex/extensions/turbo-interceptor.ts:66]: <3 thanks bro!

**Comment by gjovs** [comment]
@lucassm02 Changes dones!

**Review by lucassm02 — APPROVED** [review]
@gjovs thank you for the amazing implementation!

**Pull request merged**